daemon/emer-aggregate-tally.c

```diff
@@ -76,113 +86,132 @@ ensure_folder_exists (EmerAggregateTally  *self,
+  return swap_bytes_if_big_endian (g_variant_ref_sink (variant));
 }
 
 static guint32
-find_previous_aggregate_timer_counter (EmerAggregateTally *self,
-                                       const char         *aggregate_counter_path)
+column_to_uint32 (sqlite3_stmt *stmt,
+                  int           i)
 {
-  g_autoptr(GVariant) variant = NULL;
-  guint32 previous_counter = 0;
+  int number = sqlite3_column_int (stmt, i);
+  return *((guint32*) &number);
+}
+
+static void
+emer_aggregate_tally_constructed (GObject *object)
+{
+  EmerAggregateTally *self = EMER_AGGREGATE_TALLY (object);
+  g_autofree gchar *path = NULL;
+  int ret;
+
+  G_OBJECT_CLASS (emer_aggregate_tally_parent_class)->constructed (object);
 
-  variant = load_aggregate_timer_from_path (aggregate_counter_path);
+  g_assert (self->persistent_cache_directory != NULL);
+  ensure_folder_exists (self, self->persistent_cache_directory, NULL);
+  path = g_build_filename (self->persistent_cache_directory,
+                           "aggregate-events.db",
+                           NULL);
+  ret = sqlite3_open (path, &self->db);
+  if (ret != SQLITE_OK)
+    g_error ("Failed to open %s: %d", path, ret);
 
-  if (variant)
-    g_variant_get_child (variant, 3, "u", &previous_counter);
+  g_assert (self->db);
 
-  return previous_counter;
+  /* Use write-ahead logging rather than the default rollback journal. WAL
+   * reduces the number of writes to disk, and crucially only calls fsync()
+   * intermittently.
+   *
+   * TODO: should we force a checkpoint/fsync() to disk periodically, e.g.
+   * whenever we try to submit to the server?
+   *
+   * https://sqlite.org/wal.html
+   */
+  tally_exec_or_die (self, "PRAGMA journal_mode = WAL");
+  /* Magic number is "emer" in ASCII */
+  tally_exec_or_die (self, "PRAGMA application_id = 0x656d6572");
+  tally_exec_or_die (self, "CREATE TABLE IF NOT EXISTS tally (id INTEGER PRIMARY KEY ASC,"
+                           "                                  date TEXT NOT NULL,"
+                           "                                  event_id TEXT NOT NULL,"
+                           "                                  unix_user_id INT NOT NULL,"
+                           "                                  aggregate_key_type TEXT NOT NULL,"
+                           "                                  aggregate_key BLOB NOT NULL,"
+                           "                                  payload_type TEXT,"
+                           "                                  payload BLOB,"
+                           "                                  counter INT NOT NULL);");
+  tally_exec_or_die (self,
+                     "CREATE UNIQUE INDEX IF NOT EXISTS "
+                     "ix_tally_unique_fields ON tally (date, event_id, unix_user_id, "
+                     "                                 aggregate_key_type, aggregate_key, "
+                     "                                 payload_type, payload)");
```

Comment:
I should have added this when I wrote the draft but I think we should set:

```
  tally_exec_or_die (self, "PRAGMA user_version = 1");
```

If we never use it, it's harmless, but if we later want to use it to do versioned schema migrations it may be helpful to have.